backend/api/users.py

```diff
@@ -55,26 +58,65 @@ def create_user():
         raise ApiException(str(err), 500)
 
 
-@users.route('/api/users/', methods=['PUT'])
-# @roles_required('admin')
+@users.route('/api/users', methods=['PUT'])
+@login_required
 def update_user():
+    editor = current_user
     user = request.get_json()
-
-    res = g.user_datastore.get_user(user['email'])
-    if res is None:
-        return make_response("Unknown User with Email-address: " +
-                             user['email'], 500)
-    res.first_name = user['first name']
-    res.last_name = user['last name']
-    res.password = encrypt_password(user['password'])
-    res.bio = user['bio']
-    res.roles = []
-    res.roles.append(g.user_datastore.find_or_create_role(user['role']))
-    res.save()
-    res = make_response(dumps(res))
-    res.headers['Content-Type'] = 'application/json'
-
-    return make_response("User with email: " + user['email'] + " updated", 200)
+    is_same_user = editor['email'] == user['email']
+    if(is_permitted(editor["roles"]) is True or is_same_user is True):
+        res = g.user_datastore.get_user(user['email'])
+        if res is None:
+            return make_response("Unknown User with Email-address: " +
+                                 user['email'], 400)
+        res.first_name = user['first name']
+        res.last_name = user['last name']
+        res.bio = user['bio']
+        res.save()
+        res = make_response(dumps(res))
+        res.headers['Content-Type'] = 'application/json'
+
+        return make_response("User with email: " +
+                             user['email'] + " updated", 200)
+
+    return make_response("You don't have the permissions " +
+                         "to edit this user", 400)
+
+
+@users.route('/api/users/password', methods=['PUT'])
+@login_required
+def update_password():
+    editor = current_user
+    user = request.get_json()
+    is_same_user = editor['email'] == user['email']
+    if(is_permitted(editor["roles"]) is True and is_same_user is False):
+        res = g.user_datastore.get_user(user['email'])
+        if res is None:
+            return make_response("Unknown User with Email-address: " +
+                                 user['email'], 404)
+        new_password = user["new password"]
+        res.password = encrypt_password(new_password)
+        res.save()
+        return make_response("Password restored!", 200)
+
+    elif (editor["email"] == user['email']):
+        res = g.user_datastore.get_user(user['email'])
+        if res is None:
+            return make_response("Unknown User with Email-address: " +
+                                 user['email'], 404)
+        old_password = user["old password"]
+        if verify_password(old_password, res.password):
```

Comment:
this check should not be needed if the current user is an admin. he could as well just create a different admin user, log in with that one and then change the password of his old user. doesn't make sense, really.